### ISSUE TITLE
index out of bound in slice not caught at compilation

### ISSUE THREAD

**Issue by aybabtme** [rust-lang/rust#38035]
Hi! I'm just learning Rust! I saw this odd behavior, where the compiler will nicely tell me that an index in an array is impossible at compile time, but won't tell me that an index in a slice of known size is also impossible. 

I thought I'd share.

This fails at compile time:
```rust
let y = [1];
println!("y={}", y[1]); 
```

This fails at runtime:
```rust
let y = [1];
println!("y={}", &y[0..1][1]); 
```

Would be cool if `&y[0..1][1]` also failed at compile time.

**Comment by leonardo-m** [comment]
This may need a significant improvement of the compiler, but it's worth doing.

**Comment by clarcharr** [comment]
There are currently lints in [clippy](https://github.com/Manishearth/rust-clippy) for this. Not sure if the compiler ever plans to add these.
